crates/red_knot_python_semantic/src/types/infer.rs

```diff
@@ -4718,12 +4851,14 @@ impl<'db> TypeInferenceBuilder<'db> {
 
         for (l_ty, r_ty) in left_iter.zip(right_iter) {
             let pairwise_eq_result = self
-                .infer_binary_type_comparison(l_ty, ast::CmpOp::Eq, r_ty)
+                .infer_binary_type_comparison(l_ty, ast::CmpOp::Eq, r_ty, range)
                 .expect("infer_binary_type_comparison should never return None for `CmpOp::Eq`");
 
             match pairwise_eq_result {
                 // If propagation is required, return the result as is
                 todo @ Type::Dynamic(DynamicType::Todo(_)) => return Ok(todo),
+                // Using `bool` here is fine because we only use it to get to a truthiness
+                // but it doesn't mimic a runtime `bool` call.
```

Comment:
This comment is not right, we are mimicking a runtime `bool` call here:

```
>>> class A:
...     def __eq__(self, other):
...         return B()
...
>>> class B:
...     def __bool__(self):
...         raise RuntimeError("boom")
...
>>> (A(),) == (A(),)
Traceback (most recent call last):
  File "<python-input-16>", line 1, in <module>
    (A(),) == (A(),)
  File "<python-input-13>", line 3, in __bool__
    raise RuntimeError("boom")
RuntimeError: boom
```

Because `__eq__` is allowed to return any object, Python at runtime has to check that object for truthiness, which could fail; just like we have to do here.

(Not saying we have to fix this in this PR, just that the comment here should be a TODO instead of an incorrect rationale.)

Reply:
Nice catch. I tried to come up with a code example where we have to emit a diagnostic here but wasn't able to. Let me revisit this.

Reply:
I think the code here is actually okay. The problem is in `rich_comparison` which doesn't handle possibly unboundness or `__bool__` correctly. I'll add a test and a todo.

https://github.com/astral-sh/ruff/blob/5faf0b9c2e7c536d4a3d272a646bb77b72e41ce4/crates/red_knot_python_semantic/src/types/infer.rs#L4781-L4782